src/main/java/com/gluonhq/richtext/model/PieceTable.java

```diff
@@ -555,13 +564,92 @@ protected void doRedo(PieceTable pt) {
     public String toString() {
         return "DeleteCmd[" + deletePosition + " x " + length + "]";
     }
+
+}
+
+class ImageDecorateCmd extends AbstractCommand<PieceTable> {
+
+    private final ImageDecoration decoration;
+    private final int insertPosition;
+
+    private boolean execSuccess = false;
+    private Piece newPiece;
+    private Piece oldPiece;
+    private int opPieceIndex;
+    private Collection<Piece> newPieces = new ArrayList<>();
+
+    /**
+     * Inserts an image at the given insertion point
+     * @param decoration the image decoration
+     * @param insertPosition index of the character to decorate
+     */
+    ImageDecorateCmd(ImageDecoration decoration, int insertPosition) {
+        this.decoration = decoration;
+        this.insertPosition = insertPosition;
+    }
+
+    @Override
+    protected void doUndo(PieceTable pt) {
+        if (execSuccess) {
+            if (insertPosition == pt.getTextLength()) {
+                pt.pieces.remove(newPiece);
+                pt.fire(new TextBuffer.DeleteEvent(pt.getTextLength() - 1, 1));
+            } else {
+                pt.pieces.add(opPieceIndex, oldPiece);
+                pt.pieces.removeAll(newPieces);
+                pt.fire(new TextBuffer.DeleteEvent(insertPosition, 1));
+            }
+            pt.textLengthProperty.set(pt.getTextLength() - 1);
```

Comment:
Pressing undo after inserting an Image results in NPE:

```
Exception in thread "JavaFX Application Thread" java.lang.NullPointerException
        at com.gluonhq.richtext.model.PieceTable.walkPieces(PieceTable.java:205)
        at com.gluonhq.richtext.model.PieceCharacterIterator.reset(PieceTable.java:260)
        at com.gluonhq.richtext.model.PieceTable.resetCharacterIterator(PieceTable.java:105)
        at com.gluonhq.richtext.viewmodel.RichTextAreaViewModel.walkFragments(RichTextAreaViewModel.java:300)
        at com.gluonhq.richtext.RichTextAreaSkin.refreshTextFlow(RichTextAreaSkin.java:277)
        at com.gluonhq.richtext.RichTextAreaSkin.lambda$new$20(RichTextAreaSkin.java:128)
        at com.gluonhq.richtext.model.AbstractTextBuffer.lambda$fire$0(AbstractTextBuffer.java:33)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at com.gluonhq.richtext.model.AbstractTextBuffer.fire(AbstractTextBuffer.java:33)
        at com.gluonhq.richtext.model.ImageDecorateCmd.doUndo(PieceTable.java:600)
        at com.gluonhq.richtext.model.ImageDecorateCmd.doUndo(PieceTable.java:570)
        at com.gluonhq.richtext.undo.AbstractCommand.undo(AbstractCommand.java:12)
        at com.gluonhq.richtext.undo.CommandManager.undo(CommandManager.java:38)
        at com.gluonhq.richtext.model.PieceTable.undo(PieceTable.java:172)
        at com.gluonhq.richtext.viewmodel.RichTextAreaViewModel.undo(RichTextAreaViewModel.java:306)
        at com.gluonhq.richtext.viewmodel.RichTextAreaViewModel.undoDecoration(RichTextAreaViewModel.java:314)
        at com.gluonhq.richtext.viewmodel.DecorateCmd.doUndo(DecorateCmd.java:22)
        at com.gluonhq.richtext.viewmodel.DecorateCmd.doUndo(DecorateCmd.java:7)
        at com.gluonhq.richtext.undo.AbstractCommand.undo(AbstractCommand.java:12)
        at com.gluonhq.richtext.undo.CommandManager.undo(CommandManager.java:38)
        at com.gluonhq.richtext.viewmodel.ActionCmdUndo.apply(ActionCmdUndo.java:11)
        at com.gluonhq.richtext.action.BasicAction.execute(BasicAction.java:60)
        at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
        at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:234)
        at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
        at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
        at javafx.graphics/javafx.scene.Node.fireEvent(Node.java:8792)
        at javafx.controls/javafx.scene.control.Button.fire(Button.java:203)
        at javafx.controls/com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(ButtonBehavior.java:208)
        at javafx.controls/com.sun.javafx.scene.control.inputmap.InputMap.handle(InputMap.java:274)
        at javafx.base/com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:247)
        at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
        at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:234)
        at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
        at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
        at javafx.graphics/javafx.scene.Scene$MouseHandler.process(Scene.java:3897)
        at javafx.graphics/javafx.scene.Scene.processMouseEvent(Scene.java:1878)
        at javafx.graphics/javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2623)
        at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:411)
        at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:301)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$2(GlassViewEventHandler.java:450)
        at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:424)
        at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:449)
        at javafx.graphics/com.sun.glass.ui.View.handleMouseEvent(View.java:557)
        at javafx.graphics/com.sun.glass.ui.View.notifyMouse(View.java:943)
        at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:184)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

Reply:
Good catch, should be fixed now